client/network/src/block_request_handler.rs

```diff
@@ -61,15 +75,45 @@ pub(crate) fn generate_protocol_name(protocol_id: &ProtocolId) -> String {
 	s
 }
 
+/// The key for [`BlockRequestHandler::seen_requests`].
+#[derive(Eq, PartialEq)]
+struct SeenRequestsKey<B: BlockT> {
+	peer: PeerId,
+	from: BlockId<B>,
+	max_blocks: usize,
+	direction: Direction,
+}
+
+impl<B: BlockT> Hash for SeenRequestsKey<B> {
+	fn hash<H: Hasher>(&self, state: &mut H) {
+		self.peer.hash(state);
+		self.max_blocks.hash(state);
+		self.direction.hash(state);
+
+		match self.from {
+			BlockId::Hash(h) => h.hash(state),
+			BlockId::Number(n) => n.hash(state),
+		}
+	}
+}
+
 /// Handler for incoming block requests from a remote peer.
-pub struct BlockRequestHandler<B> {
+pub struct BlockRequestHandler<B: BlockT> {
 	client: Arc<dyn Client<B>>,
 	request_receiver: mpsc::Receiver<IncomingRequest>,
+	/// Maps from request to number of times we have seen this request.
+	///
+	/// This is used to check if a peer is spamming us with the same request.
+	seen_requests: LruCache<SeenRequestsKey<B>, usize>,
 }
 
-impl <B: BlockT> BlockRequestHandler<B> {
+impl<B: BlockT> BlockRequestHandler<B> {
 	/// Create a new [`BlockRequestHandler`].
-	pub fn new(protocol_id: &ProtocolId, client: Arc<dyn Client<B>>) -> (Self, ProtocolConfig) {
+	pub fn new(
+		protocol_id: &ProtocolId,
+		client: Arc<dyn Client<B>>,
+		network_config: &crate::config::NetworkConfiguration,
```

Comment:
```suggestion
		num_peers_hint: usize,
```

I'd prefer to not pass the entire network configuration if possible.